src/codegen/codegen_neuron_cpp_visitor.cpp

```diff
@@ -1090,9 +1147,50 @@ void CodegenNeuronCppVisitor::print_mechanism_register() {
         printer->fmt_line("pnt_receive[mech_type] = nrn_net_receive_{};", info.mod_suffix);
         printer->fmt_line("pnt_receive_size[mech_type] = {};", info.num_net_receive_parameters);
     }
+
+    if (info.thread_callback_register) {
+        printer->add_line("_nrn_thread_reg(mech_type, 1, thread_mem_init);");
+        printer->add_line("_nrn_thread_reg(mech_type, 0, thread_mem_cleanup);");
+    }
+
     printer->pop_block();
 }
 
+
+void CodegenNeuronCppVisitor::print_thread_memory_callbacks() {
+    if (!info.thread_callback_register) {
+        return;
+    }
+
+    auto static_thread_data = get_variable_name("thread_data", false);
+    auto inuse = get_variable_name("thread_data_in_use", false);
+    auto thread_data_index = info.thread_var_thread_id;
+    printer->push_block("static void thread_mem_init(Datum* _thread) ");
+    printer->push_block(fmt::format("if({})", inuse));
+    printer->fmt_line("_thread[{}] = {{neuron::container::do_not_search, new double[{}]{{}}}};",
+                      thread_data_index,
+                      info.thread_var_data_size);
+    printer->pop_block();
+    printer->push_block("else");
+    printer->fmt_line("_thread[{}] = {{neuron::container::do_not_search, {}}};",
+                      thread_data_index,
+                      static_thread_data);
+    printer->fmt_line("{} = 1;", inuse);
+    printer->pop_block();
+    printer->pop_block();
+
+    printer->push_block("static void thread_mem_cleanup(Datum* _thread) ");
+    printer->fmt_line("double * _thread_data_ptr = _thread[{}].get<double*>();", thread_data_index);
+    printer->push_block(fmt::format("if(_thread_data_ptr == {})", static_thread_data));
+    printer->fmt_line("{} = 0;", inuse);
+    printer->pop_block();
+    printer->push_block("else");
+    printer->add_line("delete[] _thread_data_ptr;");
```

Comment:
this is just a question rather than a review comment: from the old times I was wondering if we could make some of such clean-up routines redundant (i.e. replace them with smart ptrs or something).

Again, just saying if you see opportunities to change/remove old behavior, feel free to do so!

Reply:
Not immediately obvious because it needs to also be able to point to the static copy of the thread variable. It is however, another occurrence of a owning container that can act as a few.